collect_app/src/test/java/org/odk/collect/android/widgets/RangePickerDecimalWidgetTest.java

```diff
@@ -0,0 +1,147 @@
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, null));
+        int index = 0;
+
+        for (double i = 1.5F; i <= 5.5F; i += 0.5F, index++) {
+            assertThat(widget.displayedValuesForNumberPicker[index], equalTo(String.valueOf(i)));
+        }
+    }
+
+    @Test
+    public void getAnswer_whenPromptDoesNotHaveAnswer_returnsNull() {
+        assertThat(createWidget(promptWithReadOnlyAndQuestionDef(rangeQuestion)).getAnswer(), nullValue());
+    }
+
+    @Test
+    public void getAnswer_whenPromptHasAnswer_returnsAnswer() {
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, new StringData("2.5")));
+        assertThat(widget.getAnswer().getValue(), equalTo(2.5));
+    }
+
+    @Test
+    public void clearAnswer_clearsWidgetAnswer() {
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, new StringData("2.5")));
+        widget.clearAnswer();
+
+        assertThat(widget.getAnswer(), nullValue());
+        assertThat(widget.binding.widgetAnswerText.getText(), equalTo(widget.getContext().getString(R.string.no_value_selected)));
+    }
+
+    @Test
+    public void clearAnswer_callsValueChangeListener() {
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, null));
+        WidgetValueChangedListener valueChangedListener = mockValueChangedListener(widget);
+
+        widget.clearAnswer();
+        verify(valueChangedListener).widgetValueChanged(widget);
+    }
+
+    @Test
+    public void setNumberPickerValue_updatesAnswer() {
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, null));
+        widget.setNumberPickerValue(4);
+
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("3.5"));
+    }
+
+    @Test
+    public void setNumberPickerValue_whenRangeStartIsGreaterThenRangeEnd_updatesAnswer() {
+        when(rangeQuestion.getRangeStart()).thenReturn(new BigDecimal("5.5"));
+        when(rangeQuestion.getRangeEnd()).thenReturn(new BigDecimal("1.5"));
+
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, null));
+        widget.setNumberPickerValue(4);
+
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("3.5"));
+    }
+
+    @Test
+    public void clickingPickerButton_showsNumberPickerDialog() {
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, null));
+        widget.binding.widgetButton.performClick();
+
+        NumberPickerDialog numberPickerDialog = (NumberPickerDialog) widgetActivity.getActivity().getSupportFragmentManager()
+                .findFragmentByTag(NumberPickerDialog.NUMBER_PICKER_DIALOG_TAG);
+        shadowOf(getMainLooper()).idle();
+
+        assertNotNull(numberPickerDialog);
+    }
+
```

Comment:
@seadowg `clickingPickerButton_showsNumberPickerDialog()` unit test is failing. I'm looking onto it. If you have a hint, please let me know.

Reply:
Yeah if you debug you'll see that in `RangeWidgetUtils.showNumberPickerDialog` the `catch` branch will run as there is a `ClassCastException`. The code is trying to cast the `context` to a `FormEntryActivity` but it doesn't need to be one of those to add a `Fragment` right? I think it just needs to cast it to a `FragmentActivity` (`Activity` doesn't have `getSupportFragmentManager` on it).

I think you'll also need to have the activity in fully created state to add a fragment (rather than one that's just been "built").  Maybe just for those tests you can create a different test activity using `createThemedActivity`. The extra work is pretty slow so good to only use it in tests that need it!

Reply:
Right I think that's why the tests simply failed instead of giving an error